File: javaBenchmarkdf/src/test/java/benchmark/BasicMatrixBenchmark.java
package benchmark;

import basicMatrix.BasicMatrixMultiplier;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class BasicMatrixBenchmark {

    public static void main(String[] args) {
        String outputDirectory;
        if (args.length < 1) {
            outputDirectory = "data/";
            System.out.println("No output directory specified. Using default: " + outputDirectory);
        } else {
            outputDirectory = args[0];
        }

        int[] sizes = {128, 256, 512, 1024};
        int runs = 5;

        String csvPath = outputDirectory + "/java_benchmark_results.csv";

        System.out.println("==================================================");
        System.out.println("Java Matrix Multiplication Benchmark");
        System.out.println("==================================================");
        System.out.println("Matrix sizes: " + arrayToString(sizes));
        System.out.println("Runs per size: " + runs);
        System.out.println("Output directory: " + outputDirectory);

        Map<Integer, List<Double>> times = new HashMap<>();
        Map<Integer, List<Double>> realMemories = new HashMap<>();
        runBenchmarks(sizes, runs, times, realMemories, csvPath);
        printAverageResults(sizes, times, realMemories);

        System.out.println("\n==================================================");
        System.out.println("Benchmark finished. Results saved at: " + csvPath);
        System.out.println("==================================================");
    }

    private static String arrayToString(int[] array) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < array.length; i++) {
            sb.append(array[i]);
            if (i < array.length - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    private static void ensureOutputDirExists(String outputDir) {
        File dir = new File(outputDir);
        if (!dir.exists()) dir.mkdirs();
    }

    private static double getProcessMemoryMB() {
        Runtime runtime = Runtime.getRuntime();
        return (runtime.totalMemory() - runtime.freeMemory()) / (1024.0 * 1024.0);
    }

    private static void runBenchmarks(int[] sizes, int runs, Map<Integer, List<Double>> times, Map<Integer, List<Double>> realMemories, String csvPath) {
        ensureOutputDirExists(new File(csvPath).getParent());
        try (FileWriter writer = new FileWriter(csvPath)) {
            writer.write("MatrixSize,Run,TimeSeconds,RealMemoryMB\n");
            for (int size : sizes) {
                times.put(size, new ArrayList<>());
                realMemories.put(size, new ArrayList<>());
                System.out.println("\nRunning benchmarks for matrix size " + size + "x" + size + "...");

                for (int run = 1; run <= runs; run++) {
                    BasicMatrixMultiplier multiplier = new BasicMatrixMultiplier();
                    List<double[][]> matrices = multiplier.generateMatrices(size);

                    System.gc();
                    double memBefore = getProcessMemoryMB();
                    long start = System.nanoTime();
                    double[][] result = multiplier.multiply(matrices.get(0), matrices.get(1));
                    long end = System.nanoTime();
                    double memAfter = getProcessMemoryMB();

                    double timeSeconds = (end - start) / 1_000_000_000.0;
                    double realMemoryMB = Math.max(memBefore, memAfter);

                    timeSeconds = Math.round(timeSeconds * 100000.0) / 100000.0;
                    realMemoryMB = Math.round(realMemoryMB * 100000.0) / 100000.0;

                    times.get(size).add(timeSeconds);
                    realMemories.get(size).add(realMemoryMB);

                    writer.write(size + "," + run + "," + timeSeconds + "," + realMemoryMB + "\n");
                    System.out.println("  Run " + run + "/" + runs + ": " + timeSeconds + "s, " + realMemoryMB + "MB");
                }
            }
        } catch (IOException e) {
            System.out.println("Error writing benchmark results: " + e.getMessage());
        }
    }

    private static void printAverageResults(int[] sizes, Map<Integer, List<Double>> times, Map<Integer, List<Double>> realMemories) {
        System.out.println("\n===== AVERAGE RESULTS =====");
        System.out.printf("%-10s %-15s %-25s%n", "Size", "Avg Time (s)", "Avg Real Mem (MB)");
        for (int size : sizes) {
            double averageTime = times.get(size).stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
            double averageMemory = realMemories.get(size).stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
            averageTime = Math.round(averageTime * 100000.0) / 100000.0;
            averageMemory = Math.round(averageMemory * 100000.0) / 100000.0;
            System.out.printf("%-10d %-15.5f %-25.5f%n", size, averageTime, averageMemory);
        }
    }
}

File: pythonBenchmark/test/benchmark/basic_matrix_benchmark.py
import sys
import time
import csv
import os
import psutil
from basic_matrix.basic_matrix_multiplier import BasicMatrixMultiplier


sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))

def get_process_memory_mb():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / (1024 * 1024)

def run_benchmarks(sizes, runs, csv_path):
    times = {}
    real_memories = {}
    with open(csv_path, "w", newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(["MatrixSize", "Run", "TimeSeconds", "RealMemoryMB"])
        for size in sizes:
            times[size] = []
            real_memories[size] = []
            print(f"\nRunning benchmarks for matrix size {size}x{size}...")
            for run in range(1, runs + 1):
                multiplier = BasicMatrixMultiplier()
                A, B = multiplier.generate_matrices(size)
                mem_before = get_process_memory_mb()
                start = time.time()
                result = multiplier.multiply(A, B)
                end = time.time()
                mem_after = get_process_memory_mb()
                time_seconds = round(end - start, 5)
                real_memory_mb = round(max(mem_before, mem_after), 5)
                times[size].append(time_seconds)
                real_memories[size].append(real_memory_mb)
                writer.writerow([size, run, time_seconds, real_memory_mb])
                print(f"  Run {run}/{runs}: {time_seconds}s, {real_memory_mb}MB")
    return times, real_memories

def print_average_results(sizes, times, real_memories):
    print("\n===== AVERAGE RESULTS =====")
    print("{:<10} {:<15} {:<25}".format("Size", "Avg Time (s)", "Avg Real Mem (MB)"))
    for size in sizes:
        avg_time = round(sum(times[size]) / len(times[size]), 5)
        avg_real_mem = round(sum(real_memories[size]) / len(real_memories[size]), 5)
        print("{:<10} {:<15.5f} {:<25.5f}".format(size, avg_time, avg_real_mem))


if __name__ == "__main__":
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.join(script_dir, '../../..')
    output_directory = os.path.join(project_root, 'data')
    sizes = [128, 256, 512, 1024]
    runs = 5
    os.makedirs(output_directory, exist_ok=True)
    csv_path = os.path.join(output_directory, "python_benchmark_results.csv")
    print("=" * 40)
    print("Python Matrix Multiplication Benchmark")
    print("=" * 40)
    print(f"Matrix sizes: {sizes}")
    print(f"Runs per size: {runs}")
    print(f"Output directory: {output_directory}")
    times, real_memories = run_benchmarks(sizes, runs, csv_path)
    print_average_results(sizes, times, real_memories)
    print("\n" + "=" * 40)
    print(f"Benchmark finished. Results saved at: {csv_path}")
    print("=" * 40)